<a href="https://colab.research.google.com/github/hishamalboo/Data-Analysis-of-Predicting-Credit-Card-Approvals/blob/main/Predicting_Credit_Card_Approvals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 # <font size=5 color=#00b3b3>Data Analysis of Predicting Credit Card Approvals</font>


<a id='intro'></a>
<h1> Introduction</h1>

Commercial banks receive a lot of applications for credit cards. Many of them get rejected for many reasons, like high loan balances, low income levels, or too many inquiries on an individual's credit report, for example. Manually analyzing these applications is mundane, error-prone, and time-consuming (and time is money!). Luckily, this task can be automated with the power of machine learning and pretty much every commercial bank does so nowadays. In this project, you will build an automatic credit card approval predictor using machine learning techniques, just like the real banks do.

The dataset used in this project is the  <a href = "http://archive.ics.uci.edu/ml/datasets/credit+approval">"Credit Card Approval dataset"</a> from the UCI Machine Learning Repository.

We will try to cover the following tasks: 
1. Credit card applications
2. Inspecting the applications
3. Handling the missing values (part i)
4. Handling the missing values (part ii)
5. Handling the missing values (part iii)
6. Preprocessing the data (part i)
7. Splitting the dataset into train and test sets
8. Preprocessing the data (part ii)
9. Fitting a logistic regression model to the train set
10. Making predictions and evaluating performance
11. Grid searching and making the model perform better
12. Finding the best performing model

# 1. Credit card applications

Commercial banks receive a lot of applications for credit cards. Many of them get rejected for many reasons, like high loan balances, low income levels, or too many inquiries on an individual's credit report, for example. Manually analyzing these applications is mundane, error-prone, and time-consuming (and time is money!). Luckily, this task can be automated with the power of machine learning and pretty much every commercial bank does so nowadays. In this notebook, we will build an automatic credit card approval predictor using machine learning techniques, just like the real banks do.

We'll use the
The dataset used in this project is the  <a href = "http://archive.ics.uci.edu/ml/datasets/credit+approval">"Credit Card Approval dataset"</a> from the UCI Machine Learning Repository. The structure of this notebook is as follows:

First, we will start off by loading and viewing the dataset.
We will see that the dataset has a mixture of both numerical and non-numerical features, that it contains values from different ranges, plus that it contains a number of missing entries.

We will have to preprocess the dataset to ensure the machine learning model we choose can make good predictions.
After our data is in good shape, we will do some exploratory data analysis to build our intuitions.
Finally, we will build a machine learning model that can predict if an individual's application for a credit card will be accepted.
First, loading and viewing the dataset. We find that since this data is confidential, the contributor of the dataset has anonymized the feature names.

In [3]:
# Import the libraries:
from google.colab import drive
import pandas  as pd
drive.mount('/content/drive')
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns
import seaborn as sb
from sklearn import linear_model
from datetime import datetime
from timeit import default_timer as timer
from IPython.display import display
pd.set_option("display.max_columns", 10000)
pd.set_option('display.max_rows', 10000)
pd.set_option('display.width', 10000)
pd.set_option('display.max_colwidth', 1)
%load_ext google.colab.data_table


# Load dataset
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Predicting Credit Card Approvals/crx.data", header=None)

# Inspect data
df.sample(50)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
The google.colab.data_table extension is already loaded. To reload it, use:
  %reload_ext google.colab.data_table


0      1       2  3  4   5   6       7  8  9  10 11 12     13     14 15
257  b  20.00  0.000   u  g  d   v   0.500   f  f  0   f  g  00144  0      -
408  b  16.00  3.125   u  g  w   v   0.085   f  t  1   f  g  00000  6      -
576  b  30.17  0.500   u  g  c   v   1.750   t  t  11  f  g  00032  540    +
386  b  22.58  1.500   y  p  aa  v   0.540   f  f  0   t  g  00120  67     -
19   a  19.17  8.585   u  g  cc  h   0.750   t  t  7   f  g  00096  0      +
67   b  25.50  0.375   u  g  m   v   0.250   t  t  3   f  g  00260  15108  +
52   b  37.42  2.040   u  g  w   v   0.040   t  f  0   t  g  00400  5800   +
30   b  29.25  14.790  u  g  aa  v   5.040   t  t  5   t  g  00168  0      +
11   b  29.92  1.835   u  g  c   h   4.335   t  f  0   f  g  00260  200    +
247  a  19.67  0.210   u  g  q   h   0.290   t  t  11  f  g  00080  99     +
608  b  ?      0.040   y  p  d   v   4.250   f  f  0   t  g  00460  0      -
211  b  24.33  6.625   y  p  d   v   5.500   t  f  0   t  s  00100  0      +
380  b  43.17  5.000   u  g  i   bb  2.250   f  f  0   t  g  00141  0      -
303  a  15.92  2.875   u  g  q   v   0.085   f  f  0   f  g  00120  0      -
683  b  36.42  0.750   y  p  d   v   0.585   f  f  0   f  g  00240  3      -
201  a  41.33  1.000   u  g  i   bb  2.250   t  f  0   t  g  00000  300    +
300  a  57.58  2.000   u  g  ff  ff  6.500   f  t  1   f  g  00000  10     -
352  b  22.50  11.500  y  p  m   v   1.500   f  f  0   t  g  00000  4000   -
542  b  19.67  10.000  y  p  k   h   0.835   t  f  0   t  g  00140  0      -
135  b  48.58  6.500   u  g  q   h   6.000   t  f  0   t  g  00350  0      +
99   a  28.50  1.000   u  g  q   v   1.000   t  t  2   t  g  00167  500    -
258  a  20.75  9.540   u  g  i   v   0.040   f  f  0   f  g  00200  1000   -
600  a  29.50  0.460   u  g  k   v   0.540   t  t  4   f  g  00380  500    +
356  b  41.17  1.335   u  g  d   v   0.165   f  f  0   f  g  00168  0      -
7    a  22.92  11.585  u  g  cc  v   0.040   t  f  0   f  g  00080  1349   +
240  b  20.50  10.000  y  p  c   v   2.500   t  f  0   f  s  00040  0      +
371  b  26.33  13.000  u  g  e   dd  0.000   f  f  0   t  g  00140  1110   -
122  a  24.75  12.500  u  g  aa  v   1.500   t  t  12  t  g  00120  567    +
284  b  23.25  12.625  u  g  c   v   0.125   f  t  2   f  g  00000  5552   -
558  b  51.42  0.040   u  g  x   h   0.040   t  f  0   f  g  00000  3000   +
178  a  18.42  9.250   u  g  q   v   1.210   t  t  4   f  g  00060  540    +
252  a  17.83  11.000  u  g  x   h   1.000   t  t  11  f  g  00000  3000   +
133  b  36.25  5.000   u  g  c   bb  2.500   t  t  6   f  g  00000  367    +
499  b  20.42  7.000   u  g  c   v   1.625   t  t  3   f  g  00200  1391   +
58   b  35.17  4.500   u  g  x   h   5.750   f  f  0   t  s  00711  0      +
166  b  19.33  9.500   u  g  q   v   1.000   t  f  0   t  g  00060  400    +
243  a  18.75  7.500   u  g  q   v   2.710   t  t  5   f  g  ?      26726  +
130  b  67.75  5.500   u  g  e   z   13.000  t  t  1   t  g  00000  0      +
140  a  31.25  3.750   u  g  cc  h   0.625   t  t  9   t  g  00181  0      +
144  b  27.25  1.665   u  g  cc  h   5.085   t  t  9   f  g  00399  827    +
605  a  29.58  1.750   y  p  k   v   1.250   f  f  0   t  g  00280  0      -
503  b  28.25  5.125   u  g  x   v   4.750   t  t  2   f  g  00420  7      +
86   b  ?      0.375   u  g  d   v   0.875   t  f  0   t  s  00928  0      -
528  a  55.75  7.080   u  g  k   h   6.750   t  t  3   t  g  00100  50     -
307  b  18.58  10.290  u  g  ff  ff  0.415   f  f  0   f  g  00080  0      -
165  a  40.83  10.000  u  g  q   h   1.750   t  f  0   f  g  00029  837    +
623  b  18.83  0.000   u  g  q   v   0.665   f  f  0   f  g  00160  1      -
396  b  29.83  2.040   y  p  x   h   0.040   f  f  0   f  g  00128  1      -
187  b  40.58  5.000   u  g  c   v   5.000   t  t  7   f  g  00000  3065   +
49   b  23.92  0.665   u  g  c   v   0.165   f  f  0   f  g  00100  0      +

# 2. Inspecting the applications
The output may appear a bit confusing at its first sight, but let's try to figure out the most important features of a credit card application. The features of this dataset have been anonymized to protect the privacy, but this <a href = "http://rstudio-pubs-static.s3.amazonaws.com/73039_9946de135c0a49daa7a0a9eda4a67a72.html">blog</a> gives us a pretty good overview of the probable features. The probable features in a typical credit card application are Gender, Age, Debt, Married, BankCustomer, EducationLevel, Ethnicity, YearsEmployed, PriorDefault, Employed, CreditScore, DriversLicense, Citizen, ZipCode, Income and finally the ApprovalStatus. This gives us a pretty good starting point, and we can map these features with respect to the columns in the output.

As we can see from our first glance at the data, the dataset has a mixture of numerical and non-numerical features. This can be fixed with some preprocessing, but before we do that, let's learn about the dataset a bit more to see if there are other dataset issues that need to be fixed.

In [5]:
# Print summary statistics
cc_apps_description = df.describe()
print(cc_apps_description)

print("\n")

# Print DataFrame information
cc_apps_info = df.info()
print(cc_apps_info)

print("\n")

# Inspect missing values in the dataset
df.tail(17)

                2           7         10             14
count  690.000000  690.000000  690.00000  690.000000   
mean   4.758725    2.223406    2.40000    1017.385507  
std    4.978163    3.346513    4.86294    5210.102598  
min    0.000000    0.000000    0.00000    0.000000     
25%    1.000000    0.165000    0.00000    0.000000     
50%    2.750000    1.000000    0.00000    5.000000     
75%    7.207500    2.625000    3.00000    395.500000   
max    28.000000   28.500000   67.00000   100000.000000


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       690 non-null    object 
 1   1       690 non-null    object 
 2   2       690 non-null    float64
 3   3       690 non-null    object 
 4   4       690 non-null    object 
 5   5       690 non-null    object 
 6   6       690 non-null    object 
 7   7       690 non-null    float64
 8   8       690 no

0      1       2  3  4   5   6      7  8  9  10 11 12     13   14 15
673  ?  29.50  2.000   y  p  e   h   2.000  f  f  0   f  g  00256  17   -
674  a  37.33  2.500   u  g  i   h   0.210  f  f  0   f  g  00260  246  -
675  a  41.58  1.040   u  g  aa  v   0.665  f  f  0   f  g  00240  237  -
676  a  30.58  10.665  u  g  q   h   0.085  f  t  12  t  g  00129  3    -
677  b  19.42  7.250   u  g  m   v   0.040  f  t  1   f  g  00100  1    -
678  a  17.92  10.210  u  g  ff  ff  0.000  f  f  0   f  g  00000  50   -
679  a  20.08  1.250   u  g  c   v   0.000  f  f  0   f  g  00000  0    -
680  b  19.50  0.290   u  g  k   v   0.290  f  f  0   f  g  00280  364  -
681  b  27.83  1.000   y  p  d   h   3.000  f  f  0   f  g  00176  537  -
682  b  17.08  3.290   u  g  i   v   0.335  f  f  0   t  g  00140  2    -
683  b  36.42  0.750   y  p  d   v   0.585  f  f  0   f  g  00240  3    -
684  b  40.58  3.290   u  g  m   v   3.500  f  f  0   t  s  00400  0    -
685  b  21.08  10.085  y  p  e   h   1.250  f  f  0   f  g  00260  0    -
686  a  22.67  0.750   u  g  c   v   2.000  f  t  2   t  g  00200  394  -
687  a  25.25  13.500  y  p  ff  ff  2.000  f  t  1   t  g  00200  1    -
688  b  17.92  0.205   u  g  aa  v   0.040  f  f  0   f  g  00280  750  -
689  b  35.00  3.375   u  g  c   h   8.290  f  f  0   t  g  00000  0    -

# 3. Handling the missing values (part i)
We've uncovered some issues that will affect the performance of our machine learning model(s) if they go unchanged:

Our dataset contains both numeric and non-numeric data (specifically data that are of float64, int64 and object types). Specifically, the features 2, 7, 10 and 14 contain numeric values (of types float64, float64, int64 and int64 respectively) and all the other features contain non-numeric values.
The dataset also contains values from several ranges. Some features have a value range of 0 - 28, some have a range of 2 - 67, and some have a range of 1017 - 100000. Apart from these, we can get useful statistical information (like mean, max, and min) about the features that have numerical values.
Finally, the dataset has missing values, which we'll take care of in this task. The missing values in the dataset are labeled with '?', which can be seen in the last cell's output.
Now, let's temporarily replace these missing value question marks with NaN.

In [6]:
print(df.tail(17))

# Replace the '?'s with NaN
df = cc_apps.replace('?', np.nan)

# Inspect the missing values again
df.tail(17)

     0      1       2  3  4   5   6      7  8  9  10 11 12     13   14 15
673  ?  29.50  2.000   y  p  e   h   2.000  f  f  0   f  g  00256  17   -
674  a  37.33  2.500   u  g  i   h   0.210  f  f  0   f  g  00260  246  -
675  a  41.58  1.040   u  g  aa  v   0.665  f  f  0   f  g  00240  237  -
676  a  30.58  10.665  u  g  q   h   0.085  f  t  12  t  g  00129  3    -
677  b  19.42  7.250   u  g  m   v   0.040  f  t  1   f  g  00100  1    -
678  a  17.92  10.210  u  g  ff  ff  0.000  f  f  0   f  g  00000  50   -
679  a  20.08  1.250   u  g  c   v   0.000  f  f  0   f  g  00000  0    -
680  b  19.50  0.290   u  g  k   v   0.290  f  f  0   f  g  00280  364  -
681  b  27.83  1.000   y  p  d   h   3.000  f  f  0   f  g  00176  537  -
682  b  17.08  3.290   u  g  i   v   0.335  f  f  0   t  g  00140  2    -
683  b  36.42  0.750   y  p  d   v   0.585  f  f  0   f  g  00240  3    -
684  b  40.58  3.290   u  g  m   v   3.500  f  f  0   t  s  00400  0    -
685  b  21.08  10.085  y  p  e   h   1

0      1       2  3  4   5   6      7  8  9  10 11 12     13   14 15
673  NaN  29.50  2.000   y  p  e   h   2.000  f  f  0   f  g  00256  17   -
674  a    37.33  2.500   u  g  i   h   0.210  f  f  0   f  g  00260  246  -
675  a    41.58  1.040   u  g  aa  v   0.665  f  f  0   f  g  00240  237  -
676  a    30.58  10.665  u  g  q   h   0.085  f  t  12  t  g  00129  3    -
677  b    19.42  7.250   u  g  m   v   0.040  f  t  1   f  g  00100  1    -
678  a    17.92  10.210  u  g  ff  ff  0.000  f  f  0   f  g  00000  50   -
679  a    20.08  1.250   u  g  c   v   0.000  f  f  0   f  g  00000  0    -
680  b    19.50  0.290   u  g  k   v   0.290  f  f  0   f  g  00280  364  -
681  b    27.83  1.000   y  p  d   h   3.000  f  f  0   f  g  00176  537  -
682  b    17.08  3.290   u  g  i   v   0.335  f  f  0   t  g  00140  2    -
683  b    36.42  0.750   y  p  d   v   0.585  f  f  0   f  g  00240  3    -
684  b    40.58  3.290   u  g  m   v   3.500  f  f  0   t  s  00400  0    -
685  b    21.08  10.085  y  p  e   h   1.250  f  f  0   f  g  00260  0    -
686  a    22.67  0.750   u  g  c   v   2.000  f  t  2   t  g  00200  394  -
687  a    25.25  13.500  y  p  ff  ff  2.000  f  t  1   t  g  00200  1    -
688  b    17.92  0.205   u  g  aa  v   0.040  f  f  0   f  g  00280  750  -
689  b    35.00  3.375   u  g  c   h   8.290  f  f  0   t  g  00000  0    -

#5. Handling the missing values (part iii)
We have successfully taken care of the missing values present in the numeric columns. There are still some missing values to be imputed for columns 0, 1, 3, 4, 5, 6 and 13. All of these columns contain non-numeric data and this why the mean imputation strategy would not work here. This needs a different treatment.

We are going to impute these missing values with the most frequent values as present in the respective columns. 

In [7]:
# Iterate over each column of cc_apps
for col in df.columns:
    # Check if the column is of object type
    if df[col].dtypes == 'object':
        # Impute with the most frequent value
        df = df.fillna(df[col].value_counts().index[0])

# Count the number of NaNs in the dataset and print the counts to verify
print(df.isnull().sum())

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
dtype: int64


# 6. Preprocessing the data (part i)
The missing values are now successfully handled.

There is still some minor but essential data preprocessing needed before we proceed towards building our machine learning model. We are going to divide these remaining preprocessing steps into three main tasks:

Convert the non-numeric data into numeric.
Split the data into train and test sets.
Scale the feature values to a uniform range.
First, we will be converting all the non-numeric values into numeric ones. We do this because not only it results in a faster computation but also many machine learning models (like XGBoost) (and especially the ones developed using scikit-learn) require the data to be in a strictly numeric format. We will do this by using a technique called <a href = "https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html">label encoding</a>  .

In [8]:
# Import LabelEncoder
from sklearn.preprocessing import LabelEncoder

# Instantiate LabelEncoder
le=LabelEncoder()

# Iterate over all the values of each column and extract their dtypes
for col in df.columns.to_numpy():
    # Compare if the dtype is object
    if df[col].dtypes=='object':
    # Use LabelEncoder to do the numeric transformation
        df[col]=le.fit_transform(df[col])

# 7. Splitting the dataset into train and test sets
We have successfully converted all the non-numeric values to numeric ones.

Now, we will split our data into train set and test set to prepare our data for two different phases of machine learning modeling: training and testing. Ideally, no information from the test data should be used to scale the training data or should be used to direct the training process of a machine learning model. Hence, we first split the data and then apply the scaling.

Also, features like DriversLicense and ZipCode are not as important as the other features in the dataset for predicting credit card approvals. We should drop them to design our machine learning model with the best set of features. In Data Science literature, this is often referred to as feature selection.

In [9]:
# Import train_test_split
from sklearn.model_selection import train_test_split

# Drop the features 11 and 13 and convert the DataFrame to a NumPy array
df = df.drop([11, 13], axis=1)
df = df.to_numpy()

# Segregate features and labels into separate variables
X,y = df[:,0:13] , df[:,13]

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X
                                                    ,
                                                    y,
                                                    test_size=0.33,
                                                    random_state=42)

# 8. Preprocessing the data (part ii)
The data is now split into two separate sets - train and test sets respectively. We are only left with one final preprocessing step of scaling before we can fit a machine learning model to the data.

Now, let's try to understand what these scaled values mean in the real world. Let's use CreditScore as an example. The credit score of a person is their creditworthiness based on their credit history. The higher this number, the more financially trustworthy a person is considered to be. So, a CreditScore of 1 is the highest since we're rescaling all the values to the range of 0-1.

In [10]:
# Import MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

# Instantiate MinMaxScaler and use it to rescale X_train and X_test
scaler = MinMaxScaler(feature_range=(0, 1))
rescaledX_train = scaler.fit_transform(X_train)
rescaledX_test = scaler.transform(X_test)

# 9. Fitting a logistic regression model to the train set
Essentially, predicting if a credit card application will be approved or not is a classification task.<a href = "https://en.wikipedia.org/wiki/Statistical_classification">According to UCI,</a> our dataset contains more instances that correspond to "Denied" status than instances corresponding to "Approved" status. Specifically, out of 690 instances, there are 383 (55.5%) applications that got denied and 307 (44.5%) applications that got approved.

This gives us a benchmark. A good machine learning model should be able to accurately predict the status of the applications with respect to these statistics.

Which model should we pick? A question to ask is: are the features that affect the credit card approval decision process correlated with each other? Although we can measure correlation, that is outside the scope of this notebook, so we'll rely on our intuition that they indeed are correlated for now. Because of this correlation, we'll take advantage of the fact that generalized linear models perform well in these cases. Let's start our machine learning modeling with a Logistic Regression model (a generalized linear model).

In [11]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression classifier with default parameter values
logreg = LogisticRegression()

# Fit logreg to the train set
logreg.fit(rescaledX_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

# 10. Making predictions and evaluating performance
But how well does our model perform?

We will now evaluate our model on the test set with respect to <a href = "https://developers.google.com/machine-learning/crash-course/classification/accuracy">classification accuracy</a>. But we will also take a look the model's<a href = "https://www.dataschool.io/simple-guide-to-confusion-matrix-terminology/"> confusion matrix</a> . In the case of predicting credit card applications, it is equally important to see if our machine learning model is able to predict the approval status of the applications as denied that originally got denied. If our model is not performing well in this aspect, then it might end up approving the application that should have been approved. The confusion matrix helps us to view our model's performance from these aspects.

In [12]:
# Import confusion_matrix
from sklearn.metrics import confusion_matrix

# Use logreg to predict instances from the test set and store it
y_pred = logreg.predict(rescaledX_test)

# Get the accuracy score of logreg model and print it
print("Accuracy of logistic regression classifier: ", logreg.score(rescaledX_test,y_test))

# Print the confusion matrix of the logreg model
confusion_matrix(y_test,y_pred)

Accuracy of logistic regression classifier:  0.8421052631578947


array([[94,  9],
       [27, 98]])

# 11. Grid searching and making the model perform better
Our model was pretty good! It was able to yield an accuracy score of almost 84%.

For the confusion matrix, the first element of the of the first row of the confusion matrix denotes the true negatives meaning the number of negative instances (denied applications) predicted by the model correctly. And the last element of the second row of the confusion matrix denotes the true positives meaning the number of positive instances (approved applications) predicted by the model correctly.

Let's see if we can do better. We can perform <a href = "https://machinelearningmastery.com/how-to-tune-algorithm-parameters-with-scikit-learn/">  a grid search</a> of the model parameters to improve the model's ability to predict credit card approvals.

<a href = "https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html">  scikit-learn's implementation of logistic regression</a>  consists of different hyperparameters but we will grid search over the following two:

tol

max_iter

In [13]:
# Import GridSearchCV
from sklearn.model_selection import GridSearchCV

# Define the grid of values for tol and max_iter
tol = [0.01, 0.001 ,0.0001]
max_iter = [100, 150, 200]

# Create a dictionary where tol and max_iter are keys and the lists of their values are the corresponding values
param_grid = dict(tol=tol, max_iter=max_iter)

# 12. Finding the best performing model
We have defined the grid of hyperparameter values and converted them into a single dictionary format which GridSearchCV() expects as one of its parameters. Now, we will begin the grid search to see which values perform best.

We will instantiate GridSearchCV() with our earlier logreg model with all the data we have. Instead of passing train and test sets separately, we will supply X (scaled version) and y. We will also instruct GridSearchCV() to perform <a href = "https://www.dataschool.io/machine-learning-with-scikit-learn/">a cross-validation</a> of five folds.

We'll end the notebook by storing the best-achieved score and the respective best parameters.

While building this credit card predictor, we tackled some of the most widely-known preprocessing steps such as scaling, label encoding, and missing value imputation. We finished with some machine learning to predict if a person's application for a credit card would get approved or not given some information about that person.

In [14]:
# Instantiate GridSearchCV with the required parameters
grid_model = GridSearchCV(estimator=logreg, param_grid=param_grid, cv=5)

# Use scaler to rescale X and assign it to rescaledX
rescaledX = scaler.fit_transform(X)

# Fit grid_model to the data
grid_model_result = grid_model.fit(rescaledX, y)

# Summarize results
best_score, best_params = grid_model_result.best_score_, grid_model_result.best_params_
print("Best: %f using %s" % (best_score, best_params))

Best: 0.850725 using {'max_iter': 100, 'tol': 0.01}
